### PR TITLE
Mm 21672

### PR DESCRIPTION
**Pull request by 0xlilyhoffman**
#### Summary
This PR allows CompareAndSet to ignore expired values when updating database entries. Before the CompareAndSet is performed on an entry, the entry is deleted if it has expired since expired entries should be treated as if they don’t exist. 

As mentioned in https://github.com/mattermost/mattermost-server/issues/13656, there is a small window after deletIfExpired is called and before the database update is performed that a ticket may expire, resulting in incorrect behavior.  When attempting to update a value (oldValue != nil), the SQL query further guards against this case by filtering against ExpireAt. 

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/13656
https://mattermost.atlassian.net/browse/MM-21672

### CODE REVIEW

File: store/sqlstore/plugin_store.go [hunk 1]
```diff
@@ -225,6 +234,19 @@ func (ps SqlPluginStore) DeleteAllForPlugin(pluginId string) *model.AppError {
...
 
+func (ps SqlPluginStore) deleteIfExpired(pluginId, key string) *model.AppError {
+	currentTime := model.GetMillis()
+	if _, err := ps.GetMaster().Exec("DELETE FROM PluginKeyValueStore WHERE PluginId = :PluginId AND PKey = :Key AND ExpireAt != 0 AND ExpireAt < :CurrentTime",
+		map[string]interface{}{
+			"PluginId":    pluginId,
+			"Key":         key,
+			"CurrentTime": currentTime,
+		}); err != nil {
+		return model.NewAppError("SqlPluginStore.Delete", "store.sql_plugin_store.delete.app_error", nil, fmt.Sprintf("plugin_id=%v, key=%v, err=%v", pluginId, key, err.Error()), http.StatusInternalServerError)
```
Comment (lieut-data) [inline store/sqlstore/plugin_store.go]: If we don't inline, let's change the error text to be `SqlPluginStore.deleteIfExpired`,

File: store/sqlstore/plugin_store.go [hunk 2]
```diff
@@ -225,6 +234,19 @@ func (ps SqlPluginStore) DeleteAllForPlugin(pluginId string) *model.AppError {
 	return nil
 }
 
+func (ps SqlPluginStore) deleteIfExpired(pluginId, key string) *model.AppError {
```
Comment (lieut-data) [inline store/sqlstore/plugin_store.go]: I wonder if we should just inline this method for now into `CompareAndSet`?

File: store/sqlstore/plugin_store.go [hunk 3]
```diff
@@ -77,6 +77,11 @@ func (ps SqlPluginStore) CompareAndSet(kv *model.PluginKeyValue, oldValue []byte
 		return ps.CompareAndDelete(kv, oldValue)
 	}
 
+	// Expired entries should be treated as if they don't exist
```
Comment (lieut-data) [inline store/sqlstore/plugin_store.go]: Should we instead only try this if `oldValue == nil` and we intend to update? That would remove one (redundant) in lots of cases.

**Comment by jasonblais** [comment]
Hey @0xlilyhoffman, huge thanks for your contribution! Wondering if you had any questions from the feedback above? We'd be happy to clarify :)

File: store/sqlstore/plugin_store.go (hunk 1 above)
Comment (0xlilyhoffman) [reply store/sqlstore/plugin_store.go]: Personally, I'd prefer this. I'd argue it's more clear to see and understand the intentions of deleteIfExpired over another inline SQL statement. I'm open to inlining this if you feel strongly about it though.

**Comment by 0xlilyhoffman** [comment]
Hey @jasonblais thanks for checking in. My sincere apologies for the delay and silence - I was just dealing with some busy life stuff. 
I made some updates based on the feedback I've received. 
Hope that helps move this forward. 
Let me know.

**Comment by jasonblais** [comment]
No worries, thanks @0xlilyhoffman! Really appreciate it.

Looks like the build didn't pass, would you be open to taking a look? We'd be happy to help as well :)

**Comment by lieut-data** [comment]
@0xlilyhoffman, in tandem with your efforts here, I've been identifying some other issues with our key values store, and decided to rewrite the entirety of our test suite to more exhaustively cover everything. I built off your the gist of your changes, then addressed newly found issues with `List` and `KVCompareAndDelete`. I wonder if you might be willing to review https://github.com/mattermost/mattermost-server/pull/13858 in tandem with the other reviewers I've assigned?

We can figure out later how to best merge both of our efforts, but I'd love your feedback :)

**Comment by teresa-novoa** [comment]
@lieut-data #13858 is now merged, what´s the status for this on #13757? Thanks!

**Comment by lieut-data** [comment]
Thanks for the reminder, @teresa-novoa, and thanks again, @0xlilyhoffman, for your help. I'm going to close our this PR as per the changes made in my related commits.

**Pull request closed without merging**